### PR TITLE
Allow extends from empty base-class and define empty class.

### LINKED ISSUE #3320: Impossible to define icon base class for external object
It is currently not possible to define a base class with just the icon for external object classes, since the external object classes aren't allowed to contain the extends-clause needed to inherit the icon.

Could we lift the restriction on external object classes so that we don't exclude this possibility?

### PR DESCRIPTION
**Pull request by HansOlsson**
As I understand it this corresponds to the decision, but I may have missed something.
Just for external objects for now; the other part would require more complicated grammar changes.

### CODE REVIEW

File: chapters/classes.tex [hunk 1]
```diff
@@ -843,7 +843,7 @@ \subsection{Short Class Definitions}\label{short-class-definitions}
 
 \subsection{Combining Base Classes and Other Elements}\label{restriction-on-combining-base-classes-and-other-elements}\label{combining-base-classes-and-other-elements}
 
-It is not legal to combine equations, algorithms, components, base classes, or protected elements with an extends from an array class, a class with non-empty \lstinline[language=grammar]!base-prefix!, a \firstuse{simple type} (\lstinline!Real!, \lstinline!Boolean!, \lstinline!Integer!, \lstinline!String! and enumeration types), or any class transitively extending from an array class, a class with non-empty \lstinline[language=grammar]!base-prefix!, or a simple type.
+It is not legal to combine equations, algorithms, components, non-empty base classes (see \cref{empty-class}), or protected elements with an extends from an array class, a class with non-empty \lstinline[language=grammar]!base-prefix!, a \firstuse{simple type} (\lstinline!Real!, \lstinline!Boolean!, \lstinline!Integer!, \lstinline!String! and enumeration types), or any class transitively extending from an array class, a class with non-empty \lstinline[language=grammar]!base-prefix!, or a simple type.
```
Comment (henrikt-ma) [inline chapters/classes.tex:846]: Better to use definition:
```suggestion
It is not legal to combine equations, algorithms, components, non-empty base classes (see below), or protected elements with an extends from an array class, a class with non-empty \lstinline[language=grammar]!base-prefix!, a \firstuse{simple type} (\lstinline!Real!, \lstinline!Boolean!, \lstinline!Integer!, \lstinline!String! and enumeration types), or any class transitively extending from an array class, a class with non-empty \lstinline[language=grammar]!base-prefix!, or a simple type.

\begin{definition}[Empty class]\index{empty-class}
A class without equations, algorithms, or components, and where any base-classes are themselves empty.
\end{definition}

\begin{nonnormative}
An empty class may contain annotations, such as graphics, and can be used more freely as base-class than other classes.
\end{nonnormative}
```

File: chapters/classes.tex [hunk 2]
```diff
@@ -893,6 +893,11 @@ \subsection{Local Class Definitions -- Nested Classes}\label{local-class-definit
 The variables \lstinline!v1! and \lstinline!v2! are instances of this local class and thus have a nominal value of 1000.
 \end{example}
 
+\subsection{Empty Class}\label{empty-class}
+An empty class is a class without equations, algorithms, and components, and where any base-classes are themselves empty.
+
+They may contain graphics - and can be used more freely as base-classes than other classes.
+
```
Comment (henrikt-ma) [inline chapters/classes.tex:900]: Replaced by definition in suggestion above:
```suggestion
```

File: chapters/functions.tex [hunk 3]
```diff
@@ -2824,7 +2824,7 @@ \subsection{External Objects}\label{external-objects}
 \item
   An external object class shall be of the specialized class {\lstinline!class!}.
   \begin{nonnormative}
-  This is the only use of {\lstinline!class!}.
+  This is the only use of {\lstinline!class!} for a non-empty class, \cref{empty-class}.
```
Comment (henrikt-ma) [inline chapters/functions.tex:2827]: This sentence appears incomplete.  Did you mean this?
```suggestion
  Apart from empty classes (\cref{empty-class}), this is the only use of {\lstinline!class!}.
```

**Review by henrikt-ma — CHANGES_REQUESTED** [review]
In addition to suggestions, I think there are some parts missing:
- the rule for external objects should allow empty base-classes (_An external object class shall… and shall not contain other elements._)
- syntactically allow `extends enumeration(a, b, c);`
- allow combining `extends enumeration(a, b, c);` with extending from empty classes

**Comment by HansOlsson** [comment]
> the rule for external objects should allow empty base-classes (_An external object class shall… and shall not contain other elements_.)

I have tried to add that.

> syntactically allow `extends enumeration(a, b, c);`

I hadn't thought this was the idea, and to me it looks odd.
The syntactically similar:
```
type E=enumeration(a,b,c);
...
extends E;
```
should work already.

**Comment by henrikt-ma** [comment]
> > syntactically allow `extends enumeration(a, b, c);`
> 
> I hadn't thought this was the idea, and to me it looks odd.

OK.  It was my interpretation that this was the most desired alternative of the ones discussed in https://github.com/modelica/ModelicaSpecification/issues/3320#issuecomment-1470151361.  I suppose we can seek clarification of this in the next phone meeting.

In the meantime, let's try to work out some details of the alternative solution.  For example, shouldn't we support making the local type protected, so that there is only one public name for the enumeration?
```
package Icons
  type Enum
    annotation(Icon(…));
  end Enum;
end Icons;

type EnumWithIcon
protected
  type T = enumeration(a, b, c);
public
  extends T;
  extends Icons.Enum;  
end EnumWithIcon;

EnumWithIcon e = EnumWithIcon.a;
```

**Comment by HansOlsson** [comment]
> For example, shouldn't we support making the local type protected, so that there is only one public name for the enumeration?

I'm not sure. I think the current PR works for ExternalObject, and we should have a separate discussion whether more is needed for enumerations and other types.

For enumeration I would say that having a different local name has a benefit - especially as we also have to consider enumeration-connectors that should have different graphics:
```
connector TInput=input T annotation(Icon(graphics...));
connector TOutput=output T annotation(Icon(graphics...));
```
Clearly they should be the same enumeration-type.

However, if we want all enumerations-input-connectors to have a common graphics we would need a local name, and we might also consider if we can have 'input' as part of extends. On the other hand this seems like a lot of complicated grammar with somewhat unclear use-case.

**Comment by HansOlsson** [comment]
Can we accept this one (which solves the problem for ExternalObject)? And then revisit enumerations. @henrikt-ma

**Comment by henrikt-ma** [comment]
> Can we accept this one (which solves the problem for ExternalObject)? And then revisit enumerations. @henrikt-ma

Yes, just update the initial comment accordingly so that merging this doesn't trigger the closing action on #3320.

**Comment by HansOlsson** [comment]
> > Can we accept this one (which solves the problem for ExternalObject)? And then revisit enumerations. @henrikt-ma
> 
> Yes, just update the initial comment accordingly so that merging this doesn't trigger the closing action on #3320.

Done

File: chapters/functions.tex [hunk 4]
```diff
@@ -2797,7 +2797,7 @@ \subsection{External Objects}\label{external-objects}
   Since the class is partial, it is not possible to define an instance of this class.
   \end{nonnormative}
 \item
-  An external object class shall be directly extended from {\lstinline!ExternalObject!}, shall have exactly two function definitions, called {\lstinline!constructor!} and {\lstinline!destructor!}, and shall not contain other elements.
+  An external object class shall be directly extended from {\lstinline!ExternalObject!}, shall have exactly two function definitions, called {\lstinline!constructor!} and {\lstinline!destructor!}, may extend empty classes (\cref{empty-class}), but not contain any other elements.
```
Comment (henrikt-ma) [inline chapters/functions.tex:2800]: With one exception (that should also be fixed), we write say _extend from_:
```suggestion
  An external object class shall be directly extended from {\lstinline!ExternalObject!}, shall have exactly two function definitions, called {\lstinline!constructor!} and {\lstinline!destructor!}, may extend from empty classes (\cref{empty-class}), but not contain any other elements.
```

**Review by henrikt-ma — APPROVED** [review]
Just a minar comment about saying _extend from_.  This can be fixed as part of this PR, or together with the other instance of _extend_ without _from_.

**Pull request merged**